src/consensus/aft/raft.h

```diff
@@ -1426,31 +1426,35 @@ namespace aft
         }
       }
 
-      // Update next and match for the responding node.
-      auto& match_idx = node->second.match_idx;
+      // Update next or match for the responding node.
       if (r.success == AppendEntriesResponseType::FAIL)
       {
-        const auto this_match =
-          find_highest_possible_match({r.term, r.last_log_idx});
-        if (match_idx == 0)
-        {
-          match_idx = this_match;
-        }
-        else
-        {
-          match_idx = std::min(match_idx, this_match);
-        }
         // Failed due to log inconsistency. Reset sent_idx, and try again soon.
         RAFT_DEBUG_FMT(
           "Recv append entries response to {} from {}: failed",
           state->node_id,
           from);
-        node->second.sent_idx = node->second.match_idx;
+        const auto this_match =
+          find_highest_possible_match({r.term, r.last_log_idx});
+        node->second.sent_idx = std::min(this_match, node->second.sent_idx);
         return;
       }
       else
       {
-        match_idx = std::min(r.last_log_idx, state->last_idx);
+        // Potentially unnecessary safety check - use min with last_idx, to
+        // prevent matches past this node's local knowledge
+        const auto proposed_match = std::min(r.last_log_idx, state->last_idx);
+        if (proposed_match < node->second.match_idx)
+        {
+          RAFT_FAIL_FMT(
+            "Append entries response to {} from {} attempting to move "
+            "match_idx backwards ({} -> {})",
+            state->node_id,
+            from,
+            node->second.match_idx,
+            proposed_match);
```

Comment:
Am I missing something or should this function return if we get to here? Based on the comment above, it seems that we can't get here (because of `std::min()`), but if we ever did, then `match_idx` would go backwards because `node->second.match_idx = proposed_match;` is then called.

Reply:
The check above means that `r.last_log_idx` can never cause us to go backwards here, but conceivably `state->last_idx` _could_. I don't think there's ever a case where that can happen, but if it _does_ I think it's more important that we prevent `commit_idx` going past `state->last_idx` (ie - by not returning here, even if it means going backwards).

Put another way, I think we're in trouble if we ever get here, and while returning or throwing an error here are options, I think in such an unexpected situation we should err towards liveness rather than potential stalls/crashes.